### PR TITLE
Generate codeg for COUNT aggregate function

### PR DESCRIPTION
**Pull request by armenatzoglou**
In this PR we enhance GenerateAdvanceAggregates and supported built-in functions to generate code for COUNT aggregate function.

Test:
```
CREATE TABLE lineitem (
    l_orderkey integer NOT NULL,
    l_partkey integer NOT NULL,
    l_suppkey integer NOT NULL,
    l_linenumber integer NOT NULL,
    l_quantity float8 NOT NULL,
    l_extendedprice float8 NOT NULL,
    l_discount float8 NOT NULL,
    l_tax float8 NOT NULL,

    l_shipdate date NOT NULL,
    l_commitdate date NOT NULL,
    l_receiptdate date NOT NULL,
    l_returnflag char NOT NULL,
    l_linestatus char NOT NULL,

    l_shipinstruct character(25) NOT NULL,
    l_shipmode character(10) NOT NULL,
    l_comment character varying(44) NOT NULL
)
WITH (appendonly=false) DISTRIBUTED BY (l_orderkey);

\copy lineitem ( L_ORDERKEY, L_PARTKEY, L_SUPPKEY,L_LINENUMBER,L_QUANTITY, L_EXTENDEDPRICE,L_DISCOUNT,L_TAX,L_RETURNFLAG,L_LINESTATUS,L_SHIPDATE,L_COMMITDATE,L_RECEIPTDATE,L_SHIPINSTRUCT,L_SHIPMODE,L_COMMENT) from 'src/test/regress/data/lineitem.csv' with delimiter '|';
 
set codegen=on;

SELECT
   sum(l_quantity) as sum_qty,
   sum(l_extendedprice) as sum_base_price,
   sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,
   sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,
   count(*) as count_order
FROM
   lineitem
WHERE
   l_shipdate <= date '1998-12-01'
GROUP BY
   l_returnflag,
   l_linestatus
ORDER BY
   l_returnflag,
   l_linestatus;
```

### CODE REVIEW

**Comment by Eulerizeit** [comment]
Is that table from the retail demo?  

Is that a data set that is available to everyone? It should be.

**Comment by vraghavan78** [comment]
This is not retail demo but TPCH

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 1]
```diff
@@ -50,10 +49,7 @@ bool AdvanceAggregatesCodegen::GenerateAdvanceTransitionFunction(
     gpcodegen::GpCodegenUtils* codegen_utils,
     llvm::Value* llvm_pergroup_arg,
     int aggno,
-    llvm::Function* advance_aggregates_func,
-    llvm::BasicBlock* error_block,
-    llvm::Value *llvm_in_arg_ptr) {
-
+    gpcodegen::PGFuncGeneratorInfo& pg_func_info) {
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: +1 for this change

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 2]
```diff
@@ -88,19 +84,16 @@ bool AdvanceAggregatesCodegen::GenerateAdvanceTransitionFunction(
...
 
-  // FunctionCallInvoke(fcinfo); {{
-  // We do not need to use a FunctionCallInfoData struct, since the supported
-  // aggregate functions are simple enough.
-  gpcodegen::PGFuncGeneratorInfo pg_func_info(
-      advance_aggregates_func,
-      error_block,
-      {irb->CreateLoad(llvm_pergroupstate_transValue_ptr),
-          irb->CreateLoad(llvm_in_arg_ptr)});
+  assert(nullptr != peraggstate->aggref &&
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Minor : Does it need to be in the same assert ?

File: src/backend/codegen/op_expr_tree_generator.cc [hunk 3]
```diff
@@ -93,6 +93,18 @@ void OpExprTreeGenerator::InitializeSupportedFunction() {
           "int8pl",
           &PGArithFuncGenerator<int64_t, int64_t, int64_t>::AddWithOverflow));
 
+  supported_function_[1219] = std::unique_ptr<PGFuncGeneratorInterface>(
```
Comment (karthijrk) [inline src/backend/codegen/op_expr_tree_generator.cc]: oh, we have two opcode for same function signature ?

Comment (armenatzoglou) [inline src/backend/codegen/op_expr_tree_generator.cc]: Yes.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 2 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: No. I will split them.

**Comment by armenatzoglou** [comment]
Pushed 020b71a56d82ead1416bad672b428d8617355f1d. Thanks a lot for the review.

**Pull request closed without merging**